### PR TITLE
Dart improvements

### PR DESCRIPTION
**Pull request by robertbastian**
Better slices, equality, function type heuristics

### CODE REVIEW

File: tool/templates/dart/slice.dart.jinja [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+      {% endif -%}
+        return false;
+      }
+    }
+    return true;
+  }
+
+  // This is cheap
+  @override
+  int get hashCode => _length.hashCode;
```
Comment (Manishearth) [inline tool/templates/dart/slice.dart.jinja:43]: question: is it correct for the hashcode of a slice to just be based on its length?

File: tool/src/dart/class.rs [hunk 2]
```diff
@@ -363,23 +364,62 @@ impl<'a, 'dartcx, 'tcx: 'dartcx> TyGenContext<'a, 'dartcx, 'tcx> {
...
+                )
             } else {
                 format!("static {return_ty} {method_name}({params})")
             }
-        // } else if method.params.is_empty() && return_ty != "void" && method_name != "toString" {
-        //     format!("{return_ty} get {method_name}")
-        // } else if method_name.starts_with("set") && method.params.len() == 1 {
-        //     use heck::ToLowerCamelCase;
-        //     format!("{return_ty} set {}({params})", method_name.strip_prefix("set").unwrap().to_lower_camel_case())
+        } else if method.params.is_empty()
```
Comment (Manishearth) [inline tool/src/dart/class.rs:388]: nit: I think this logic belongs in the `fmt()` family of functions

File: tool/templates/dart/slice.dart.jinja (hunk 1 above)
Comment (robertbastian) [reply tool/templates/dart/slice.dart.jinja:43]: yes, Dart uses the standard hashcode contract: https://api.dart.dev/stable/3.1.5/dart-core/Object/hashCode.html

See https://github.com/unicode-org/icu4x/pull/4244#discussion_r1383019404

File: tool/src/dart/class.rs (hunk 2 above)
Comment (robertbastian) [reply tool/src/dart/class.rs:388]: I disagree, this is pretty fundamental to what the method is, it's not just a cosmetic change.

**Review by Manishearth — APPROVED** [review]

**Review by Manishearth — APPROVED** [review]

**Pull request merged**